etl/steps/data/garden/migration/2024-11-19/migration_distance.py

```diff
@@ -0,0 +1,131 @@
+"""Load a meadow dataset and create a garden dataset."""
+
+import geopandas as gpd
+import pandas as pd
+from owid.catalog import License, Origin
+
+from etl.data_helpers import geo
+from etl.helpers import PathFinder, create_dataset
+
+# Get paths and naming conventions for current step.
+paths = PathFinder(__file__)
+
+# GeoJson file with country boundaries
+URL_TINY_CTY = "https://d2ad6b4ur7yvpq.cloudfront.net/naturalearth-3.3.0/ne_50m_admin_0_tiny_countries.geojson"
+
+NATURAL_EARTH_ORIGIN = Origin(
+    producer="Natural Earth",
+    title="Admin 0 - Countries and Tiny Country Points",
+    url_main="https://www.naturalearthdata.com/",
+    url_download="https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/110m/cultural/ne_110m_admin_0_countries.zip",
+    description="""Natural Earth is a public domain map dataset available at 1:10m, 1:50m, and 1:110 million scales. It features tightly integrated vector and raster data, was built through a collaboration of many volunteers and is supported by NACIS (North American Cartographic Information Society).""",
+    license=License(name="Public Domain", url="https://www.naturalearthdata.com/about/terms-of-use/"),
+    date_accessed="2024-11-19",
+    citation_full="Made with Natural Earth. Free vector and raster map data @ naturalearthdata.com.",
+)
+
+
+def run(dest_dir: str) -> None:
+    #
+    # Load inputs.
+    #
+    # Load meadow dataset.
+    ds_meadow = paths.load_dataset("migrant_stock")
+
+    # Read table from meadow dataset.
+    tb = ds_meadow["migrant_stock_dest_origin"].reset_index()
+
+    # read natural earth data
+    world = gpd.read_file(gpd.datasets.get_path("naturalearth_lowres"))  # type: ignore
+    tiny_countries = gpd.read_file(URL_TINY_CTY)
+
+    # use web mercator projection
+    world = world.to_crs("EPSG:3857")[["name", "geometry"]]
+    tiny_countries = tiny_countries.to_crs("EPSG:3857")[["admin", "geometry"]]
+    tiny_countries = tiny_countries.rename(columns={"admin": "name"})
+
+    # add tiny countries to world
+    world = pd.concat([world, tiny_countries], ignore_index=True)
+
+    # harmonize country names
+    world = geo.harmonize_countries(
+        df=world,  # type: ignore
+        country_col="name",
+        countries_file=paths.country_mapping_path,
+    )
+
+    # Calculate distance matrix (in km)
+    distance_matrix = (
+        world.geometry.apply(lambda geom1: world.geometry.apply(lambda geom2: geom1.distance(geom2))) / 1e3
+    )
+
+    # Add country names to the distance matrix
+    distance_df = distance_matrix.copy()
+    distance_df.index = world["name"]
+    distance_df.columns = world["name"]
+
+    ## Add distances to migration flows table
+    # Remove "Other" and countries without distance data from country destination or country origin columns
+    cty_no_data = [
```

Comment:
These are missing from the UN DESA data?

Reply:
These are in the UN DESA but missing from the natural earth data! Since they have very small migration flows I've decided to exclude them instead of trying to hunt down another data set.

Reply:
Ah interesting, it may be because you were using the low res natural earth data, which has much coarser boundaries and because of this might skip small countries. I'm not sure what version Lucas imported, but it might be different.

Reply:
But I think it is also totally reasonable to skip them for the reason you give!